MQTT listener does not restart subscription on reconnection

The MQTT library we are using is resilient to dropped TCP connections and will reconnect  unfortunately the it does not restart the subscriptions once reconnected.

I think we need to either -
- keep track of running subscriptions and restart within the listener
- tell the runtime/manager to invalidate all of the running pipes and try to restart them.
- something else

I prefer the second option as it sounds more generic and applicable to other use cases.

Last night I got this, we have to prevent a panic like this in the server.
```
Christians-MBP:darwin-amd64 christian[client-config-file !?]$ ./device-hub server
2017/07/17 18:59:53 Listening on: [:50051]
2017/07/17 19:04:13 transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:50051->127.0.0.1:59198: read: connection reset by peer
err not found
2017/07/17 22:37:44 mqtt broker connected
2017/07/18 00:56:44 mqtt broker disconnectedpingresp not received, disconnecting
panic: mqtt broker disconnectedpingresp not received, disconnecting

goroutine 1150 [running]:
log.Panic(0xc420031f90, 0x2, 0x2)
	/usr/local/go/src/log/log.go:322 +0xc0
github.com/thingful/device-hub/listener.Register.func1.1(0x1a527a0, 0xc420077180, 0x1a45340, 0xc420202020)
	/Users/christian/go/src/github.com/thingful/device-hub/listener/listeners.go:71 +0xd2
created by github.com/thingful/device-hub/vendor/github.com/eclipse/paho%2emqtt%2egolang.(*client).internalConnLost
	/Users/christian/go/src/github.com/thingful/device-hub/vendor/github.com/eclipse/paho.mqtt.golang/client.go:418 +0x109
```

--- a/listener/listeners.go
+++ b/listener/listeners.go
@@ -68,7 +68,7 @@ func Register(r *registry.Registry) {
 				log.Print("mqtt broker connected")
 			}
 
-			// Now this codeis weird but let me explain...
+			// Now this code is weird but let me explain...
 			// The call to mqtt.NewClient takes a pointer to an mqtt.Options however
 			// dereferences it before capturing a copy of the struct which while
 			// quite neat really is a pain when wiring up the OnConnectionLost callback

--- a/listener/listeners.go
+++ b/listener/listeners.go
@@ -5,6 +5,7 @@ package listener
 import (
 	"fmt"
 	"log"
+	"os"
 	"time"
 
 	mqtt "github.com/eclipse/paho.mqtt.golang"
@@ -41,6 +42,8 @@ func Register(r *registry.Registry) {
 	r.RegisterListener("mqtt",
 
 		func(config describe.Values) (hub.Listener, error) {
+			mqtt.DEBUG = log.New(os.Stdout, "", 0)
+			mqtt.ERROR = log.New(os.Stdout, "", 0)
 
 			brokerAddress := config.MustString(mqtt_bindingAddress.Name)
 			clientID := fmt.Sprintf("device-hub-%s", hub.SourceVersion)
@@ -64,17 +67,6 @@ func Register(r *registry.Registry) {
 			opts.SetPingTimeout(10 * time.Second)
 			opts.SetAutoReconnect(true)
 
-			// Panic on connection lost until
-			// https://github.com/thingful/device-hub/issues/27
-			// is resolved
-			opts.OnConnectionLost = func(client mqtt.Client, err error) {
-				log.Panic("mqtt broker disconnected", err)
-			}
-
-			opts.OnConnect = func(mqtt.Client) {
-				log.Print("mqtt broker connected")
-			}
-
 			client := mqtt.NewClient(opts)
 
 			// TODO : set sensible wait time
@@ -82,7 +74,29 @@ func Register(r *registry.Registry) {
 				return nil, token.Error()
 			}
 
-			return newMQTTListener(client)
+			listener, err := newMQTTListener(client)
+
+			if err != nil {
+				return nil, err
+			}
+
+			opts.OnConnectionLost = func(client mqtt.Client, err error) {
+
+				log.Println("mqtt broker disconnected", err)
+				log.Println("attempting to reconnect existing subscriptions")
+
+				err2 := listener.RestartSubscriptions()
+
+				if err2 != nil {
+					log.Panicf("error restarting subscriptions - %v", err2)
+				}
+
+			}
+
+			opts.OnConnect = func(mqtt.Client) {
+				log.Print("mqtt broker connected")
+			}
+			return listener, nil
 
 		},
 		describe.Parameters{

--- a/listener/listeners.go
+++ b/listener/listeners.go
@@ -5,7 +5,6 @@ package listener
 import (
 	"fmt"
 	"log"
-	"os"
 	"time"
 
 	mqtt "github.com/eclipse/paho.mqtt.golang"
@@ -42,8 +41,6 @@ func Register(r *registry.Registry) {
 	r.RegisterListener("mqtt",
 
 		func(config describe.Values) (hub.Listener, error) {
-			mqtt.DEBUG = log.New(os.Stdout, "", 0)
-			mqtt.ERROR = log.New(os.Stdout, "", 0)
 
 			brokerAddress := config.MustString(mqtt_bindingAddress.Name)
 			clientID := fmt.Sprintf("device-hub-%s", hub.SourceVersion)
@@ -67,21 +64,21 @@ func Register(r *registry.Registry) {
 			opts.SetPingTimeout(10 * time.Second)
 			opts.SetAutoReconnect(true)
 
-			client := mqtt.NewClient(opts)
-
-			// TODO : set sensible wait time
-			if token := client.Connect(); token.Wait() && token.Error() != nil {
-				return nil, token.Error()
+			opts.OnConnect = func(mqtt.Client) {
+				log.Print("mqtt broker connected")
 			}
 
-			listener, err := newMQTTListener(client)
-
-			if err != nil {
-				return nil, err
-			}
+			// Now this codeis weird but let me explain...
+			// The call to mqtt.NewClient takes a pointer to an mqtt.Options however
+			// dereferences it before capturing a copy of the struct which while
+			// quite neat really is a pain when wiring up the OnConnectionLost callback
+			// that operates on an instance of a mqttlistener
+			// So... we create a reference to an mqttl istener, wire it up then instantiate
+			// it which is weird but saves creating another layer of indirection
+			var listener *mqttlistener
+			var err error
 
 			opts.OnConnectionLost = func(client mqtt.Client, err error) {
-
 				log.Println("mqtt broker disconnected", err)
 				log.Println("attempting to reconnect existing subscriptions")
 
@@ -90,12 +87,21 @@ func Register(r *registry.Registry) {
 				if err2 != nil {
 					log.Panicf("error restarting subscriptions - %v", err2)
 				}
-
 			}
 
-			opts.OnConnect = func(mqtt.Client) {
-				log.Print("mqtt broker connected")
+			client := mqtt.NewClient(opts)
+
+			// TODO : set sensible wait time
+			if token := client.Connect(); token.Wait() && token.Error() != nil {
+				return nil, token.Error()
 			}
+
+			listener, err = newMQTTListener(client)
+
+			if err != nil {
+				return nil, err
+			}
+
 			return listener, nil
 
 		},

--- a/listener/mqtt.go
+++ b/listener/mqtt.go
@@ -4,6 +4,7 @@ package listener
 
 import (
 	"errors"
+	"sync"
 
 	mqtt "github.com/eclipse/paho.mqtt.golang"
 	hub "github.com/thingful/device-hub"
@@ -16,18 +17,31 @@ func newMQTTListener(client mqtt.Client) (*mqttlistener, error) {
 	}
 
 	return &mqttlistener{
-		client: client,
+		client:        client,
+		subscriptions: map[string]defaultChannel{},
 	}, nil
 }
 
 type mqttlistener struct {
 	client mqtt.Client
+
+	lock          sync.RWMutex
+	subscriptions map[string]defaultChannel
 }
 
 func (m *mqttlistener) NewChannel(topic string) (hub.Channel, error) {
 
 	if topic == "" {
 		return nil, errors.New("mqtt topic is empty string")
+	}
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	_, found := m.subscriptions[topic]
+
+	if found {
+		return nil, errors.New("unable to start subscription for existing topic")
 	}
 
 	errors := make(chan error)
@@ -42,15 +56,39 @@ func (m *mqttlistener) NewChannel(topic string) (hub.Channel, error) {
 		return nil, token.Error()
 	}
 
-	return defaultChannel{out: out, errors: errors, close: func() error {
+	channel := defaultChannel{out: out, errors: errors, close: func() error {
 		token := m.client.Unsubscribe(topic)
 		token.Wait()
 		return token.Error()
-	}}, nil
+	}}
+
+	m.subscriptions[topic] = channel
+
+	return channel, nil
 }
 
 func (m *mqttlistener) Close() error {
 	// TODO : set a sensible timeout
 	m.client.Disconnect(1000)
+	return nil
+}
+
+func (m *mqttlistener) RestartSubscriptions() error {
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	for topic, s := range m.subscriptions {
+
+		handler := func(client mqtt.Client, msg mqtt.Message) {
+			input := newHubMessage(msg.Payload(), "MQTT", msg.Topic())
+			s.out <- input
+		}
+
+		if token := m.client.Subscribe(topic, 0, handler); token.Wait() && token.Error() != nil {
+			return token.Error()
+		}
+	}
+
 	return nil
 }

